⚠️ add docker machine pool and e2e tests for machine pool

devigned opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
Currently, the test framework is unaware of machine pools and will not wait for those nodes to be ready. This PR is to enhance the test framework to include machine pool functionality.

This PR adds a `DockerMachinePool` implementation. As part of adding that implementation, the PR also expands upon the e2e tests to validate the `DockerMachinePool`.

devigned commented:
/retest

devigned commented:
api-diff
```
sigs.k8s.io/cluster-api/test/framework/clusterctl
  Incompatible changes:
  - ApplyClusterTemplateAndWait: changed from func(context.Context, ApplyClusterTemplateAndWaitInput) (*sigs.k8s.io/cluster-api/api/v1alpha3.Cluster, *sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3.KubeadmControlPlane, []*sigs.k8s.io/cluster-api/api/v1alpha3.MachineDeployment) to func(context.Context, ApplyClusterTemplateAndWaitInput) *ApplyClusterTemplateAndWaitResult
  Compatible changes:
  - ApplyClusterTemplateAndWaitInput.WaitForMachinePools: added
  - ApplyClusterTemplateAndWaitResult: added

sigs.k8s.io/cluster-api/test/framework
  Incompatible changes:
  - GetMachineDeploymentsByClusterInput: removed
  Compatible changes:
  - DiscoveryAndWaitForMachinePools: added
  - DiscoveryAndWaitForMachinePoolsInput: added
  - GetMachinePoolsByCluster: added
  - GetResourcesByClusterInput: added
  - WaitForMachinePoolNodesToExist: added
  - WaitForMachinePoolNodesToExistInput: added
```

devigned commented:
@fabriziopandini what would you think about me adding docker infra implementation of machine pool?

fabriziopandini commented:
> @fabriziopandini what would you think about me adding docker infra implementation of machine pool?

Sorry for lagging in the response.
I'm +1 to this, given that this will allow validation of machine pools in the CAPI e2e
I available to help if we agree on high-level design and there is a chance to break down the work

fabriziopandini commented on test/infrastructure/docker/exp/api/v1alpha3/dockermachinepool_types.go:
```diff
@@ -0,0 +1,105 @@
+type DockerMachinePoolSpec struct {
+	// Template contains the details used to build a replica machine within the Machine Pool
+	// +optional
+	Template DockerMachineTemplate `json:"template"`
```
Might be I'm missing some context here, but I find a little bit confusing having a  `DockerMachineTemplate` type different than the one defined in the main CAPD API.

Could we use `DockerMachineSpec` (similar to what we are doing for the template in `MachinePoolSpec` https://github.com/kubernetes-sigs/cluster-api/blob/cf467a0fefea032290e3261ee89d98e3d7623ca5/exp/api/v1alpha3/machinepool_types.go#L44)

devigned replied:
Not all of the fields in `DockerMachineTemplate` were applicable for the machine pool template. For example, `Bootstrapped` doesn't make sense on a template. `Boostrapped` makes sense on the status of an instance of a template, but not on the template itself. Also, it seems like `Bootstrapped` would be better placed on the status rather than the spec, but that's loosely related.

fabriziopandini commented on test/infrastructure/docker/exp/README.md:
```diff
@@ -0,0 +1,23 @@
+## Active Features
+ MachinePool  (alpha)
+
+ ClusterResourceSet (alpha)
```
```suggestion
```

fabriziopandini commented on test/infrastructure/docker/exp/README.md:
```diff
@@ -0,0 +1,23 @@
+For more information on graduation criteria, see: [Contributing Guidelines](../CONTRIBUTING.md#experiments)
+
+## Active Features
+ MachinePool  (alpha)
```
```suggestion
 DockerMachinePool  (alpha)
```

fabriziopandini commented on test/infrastructure/docker/examples/machine-pool.yaml:
```diff
@@ -0,0 +1,104 @@
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: DockerMachinePool
+metadata:
+  name: worker-0
```
Should this be
```suggestion
  name: worker-dmp-0
```

fabriziopandini commented on test/framework/clusterctl/clusterctl_helpers.go:
```diff
@@ -109,11 +110,19 @@ type ApplyClusterTemplateAndWaitInput struct {
 // ApplyClusterTemplateAndWait gets a cluster template using clusterctl, and waits for the cluster to be ready.
 // Important! this method assumes the cluster uses a KubeadmControlPlane and MachineDeployments.
-func ApplyClusterTemplateAndWait(ctx context.Context, input ApplyClusterTemplateAndWaitInput) (*clusterv1.Cluster, *controlplanev1.KubeadmControlPlane, []*clusterv1.MachineDeployment) {
+func ApplyClusterTemplateAndWait(ctx context.Context, input ApplyClusterTemplateAndWaitInput) *ApplyClusterTemplateAndWaitResult {
```
This is a breaking change for all the consumers of the test framework, so we should document an action required in the release note or wait for v1alpha4
@vincepri ^^

fabriziopandini commented on test/e2e/kcp_upgrade.go:
```diff
@@ -50,8 +49,7 @@ func KCPUpgradeSpec(ctx context.Context, inputGetter func() KCPUpgradeSpecInput)
 		cancelWatches context.CancelFunc
-		cluster       *clusterv1.Cluster
-		controlPlane  *controlplanev1.KubeadmControlPlane
+		result        *clusterctl.ApplyClusterTemplateAndWaitResult
```
What about?
```suggestion
		clusterReosurces        *clusterctl.ApplyClusterTemplateAndWaitResult
```
(result does not express what the variables really is)

fabriziopandini commented on test/framework/machinedeployment_helpers.go:
```diff
@@ -55,8 +56,8 @@ func CreateMachineDeployment(ctx context.Context, input CreateMachineDeploymentI
-// GetMachineDeploymentsByClusterInput is the input for GetMachineDeploymentsByCluster.
-type GetMachineDeploymentsByClusterInput struct {
+// GetResourcesByClusterInput is the input for GetMachineDeploymentsByCluster and GetMachinePoolsByCluster.
+type GetResourcesByClusterInput struct {
```
This is another breaking change.
However, in this case, IMO we should keep the current type and create a new one for `GetMachinePoolsByCluster`, so we are not coupling the two methods

devigned replied:
Kept the current type and added a new one for `MachinePools`

fabriziopandini commented on test/framework/machinedeployment_helpers.go:
```diff
@@ -76,6 +77,20 @@ func GetMachineDeploymentsByCluster(ctx context.Context, input GetMachineDeploym
+// GetMachinePoolsByCluster returns the MachinePools objects for a cluster.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMachinePoolsByCluster(ctx context.Context, input GetResourcesByClusterInput) []*clusterv1exp.MachinePool {
```
Should this goes in a separated file named `machinedpool_helpers.go` (same for other new types/func below)

devigned replied:
As suggested, I moved the `MachinePool` functionality to a new file.

fabriziopandini reviewed (commented):
First pass. I'm really happy to see this happening!
/cc @srm09

devigned commented:
@fabriziopandini, I've incorporated your feedback. I've also added an implementation for the `DockerMachinePool` reconciler (sans conditions). Please take a look. I'd appreciate your feedback on the design. Thank you.

Updated api diff:
```
sigs.k8s.io/cluster-api/exp/util
  Compatible changes:
  - MachinePoolToInfrastructureMapFunc: added

sigs.k8s.io/cluster-api/test/framework/clusterctl
  Incompatible changes:
  - ApplyClusterTemplateAndWait: changed from func(context.Context, ApplyClusterTemplateAndWaitInput) (*sigs.k8s.io/cluster-api/api/v1alpha3.Cluster, *sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3.KubeadmControlPlane, []*sigs.k8s.io/cluster-api/api/v1alpha3.MachineDeployment) to func(context.Context, ApplyClusterTemplateAndWaitInput) *ApplyClusterTemplateAndWaitResult
  Compatible changes:
  - ApplyClusterTemplateAndWaitInput.WaitForMachinePools: added
  - ApplyClusterTemplateAndWaitResult: added

sigs.k8s.io/cluster-api/test/framework
  Compatible changes:
  - DiscoveryAndWaitForMachinePools: added
  - DiscoveryAndWaitForMachinePoolsInput: added
  - GetMachinePoolsByCluster: added
  - GetMachinePoolsByClusterInput: added
  - WaitForMachinePoolNodesToExist: added
  - WaitForMachinePoolNodesToExistInput: added
```

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+			},
+		).
+		Watches(
+			&source.Kind{Type: &infrav1.DockerCluster{}},
```
Question: Why are we watching for changes to DockerCluster?

devigned replied:
No good reason. Removed.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// check if there are more machines than the spec requires
+	realMachineCount := int32(len(dockerMachinePool.Status.Machines))
+	if realMachineCount > 0 && realMachineCount > *machinePool.Spec.Replicas {
```
In the case of delete, I was expecting the controller to scale down to 0, while instead, it is scaling down to `Spec.Replicas`. Am I missing something?

devigned replied:
Delete now removes all docker machines labeled with the machine pool.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// check if there are more machines than the spec requires
+	realMachineCount := int32(len(dockerMachinePool.Status.Machines))
```
What about getting the real machine count from docker, instead of relying on status (which reports the situation at the latest reconciliation, that might be different from the current one)

devigned replied:
This has been updated as a "nodePool" which is driven from the state of docker rather than the status of the machine pool

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+		return ctrl.Result{}, nil
+	}
+
+	if machinePool.Spec.Replicas == nil {
```
Move the defaulting web hook?

devigned replied:
I was hesitant to add more to this monster of a PR.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+	realMachineCount := int32(len(dockerMachinePool.Status.Machines))
+	if realMachineCount > 0 && realMachineCount > *machinePool.Spec.Replicas {
+		for i := *machinePool.Spec.Replicas; i < realMachineCount; i++ {
+			if err := r.reconcileDockerWorkerMachine(ctx, cluster.Name, i, machinePool, dockerMachinePool, log); err != nil {
```
Should this be deleteDockerWorkerMachine?

devigned replied:
Yes. Fixed.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+	}
+
+	needToRequeue := false
+	for i := int32(0); i < *machinePool.Spec.Replicas; i++ {
```
I think that we should always read the real state from docker, and then decide if to create or to delete machine, e.g.
- list the current real machines (that are the containers controlled by the MachinePool, detected by naming convention or using labels)
- reconcile machines that should exist
- delete machines that should not exist

devigned replied:
Agreed, and updated to reflect this feedback.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+	// Usually a cloud provider will do this, but there is no docker-cloud provider.
+	// Requeue if there is an error, as this is likely momentary load balancer
+	// state changes during control plane provisioning.
+	if err := externalMachine.SetNodeProviderID(ctx); err != nil {
```
Wrap with if ProviderID is not set?

devigned replied:
+1

fabriziopandini reviewed (commented):
@devigned this is a big jump forward!
I think that we can make the reconcile loop more robust if we start from a list of existing containers and then adapt the current real state to desired.
Also, it will be great if we can find a way to implement once the bits that are identical from MachinePool and Machine (e.g. getBootstrapToken, SetProviderID), but at the same time I don't want to expand too much the scope of this PR, so I leave this evaluation to you.

devigned commented on test/infrastructure/docker/exp/api/v1alpha3/dockermachinepool_types.go:
```diff
@@ -0,0 +1,132 @@
+	// +optional
+	ProviderID string `json:"providerID,omitempty"`
+
+	// ProviderIDList are the identification IDs of Machine Pool machineq instances
```
InstanceID has since been replaced with the instance name used for identifying the docker machine related to the machine status structure.

devigned commented on test/infrastructure/docker/exp/api/v1alpha3/dockermachinepool_types.go:
```diff
@@ -0,0 +1,132 @@
+
+	// ProviderID is the provider identification of the Machine Pool Instance
+	// +optional
+	ProviderID *string `json:"providerID,omitempty"`
```
Perhaps, not, but it is convenient for capturing the providerID so it can be used higher up the stack.

devigned commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,459 @@
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Fetch the Docker Cluster.
```
removed

devigned commented on test/infrastructure/docker/exp/api/v1alpha3/dockermachinepool_types.go:
```diff
@@ -0,0 +1,132 @@
+
+	// Machines contains the status for each machine in the pool
+	// +optional
+	Machines []DockerMachinePoolInstanceStatus `json:"machines,omitempty"`
```
Changed to `Instances`.

devigned commented:
@fabriziopandini, I think I've addressed your feedback, and have added some e2e tests.

Updated API Diff
```

sigs.k8s.io/cluster-api/test/framework
  Compatible changes:
  - DiscoveryAndWaitForMachinePools: added
  - DiscoveryAndWaitForMachinePoolsInput: added
  - GetMachinePoolInstanceVersions: added
  - GetMachinePoolsByCluster: added
  - GetMachinePoolsByClusterInput: added
  - GetMachinesPoolInstancesInput: added
  - ScaleMachinePoolAndWait: added
  - ScaleMachinePoolAndWaitInput: added
  - UpgradeMachinePoolAndWait: added
  - UpgradeMachinePoolAndWaitInput: added
  - WaitForMachinePoolInstancesToBeUpgraded: added
  - WaitForMachinePoolInstancesToBeUpgradedInput: added
  - WaitForMachinePoolNodesToExist: added
  - WaitForMachinePoolNodesToExistInput: added

sigs.k8s.io/cluster-api/test/framework/clusterctl
  Incompatible changes:
  - ApplyClusterTemplateAndWait: changed from func(context.Context, ApplyClusterTemplateAndWaitInput) (*sigs.k8s.io/cluster-api/api/v1alpha3.Cluster, *sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3.KubeadmControlPlane, []*sigs.k8s.io/cluster-api/api/v1alpha3.MachineDeployment) to func(context.Context, ApplyClusterTemplateAndWaitInput) *ApplyClusterTemplateAndWaitResult
  Compatible changes:
  - (*E2EConfig).GetInt32PtrVariable: added
  - ApplyClusterTemplateAndWaitInput.WaitForMachinePools: added
  - ApplyClusterTemplateAndWaitResult: added

sigs.k8s.io/cluster-api/exp/util
  Compatible changes:
```

devigned commented:
/hold

I'm noticing some flakes locally. I'm going to investigate.

devigned commented:
/hold cancel

Adding a bit more time for machine pools to reconcile seems to remedy the flakes.

devigned commented:
/test pull-cluster-api-test

devigned commented:
/test pull-cluster-api-e2e-full

devigned commented:
/test pull-cluster-api-e2e-full

Updated with latest to take advantage of #3667 since self-hosted was failing in the full e2e tests.

fabriziopandini reviewed (commented):
First pass.
I still need to carve out some time to review the controller

devigned commented on test/infrastructure/docker/docker/types/node.go:
```diff
@@ -34,13 +34,15 @@ type Node struct {
 
 // NewNode returns a Node with defaults.
-func NewNode(name, role string) *Node {
+func NewNode(name, image, role string) *Node {
```
It is used to feed the image the container was created with up the stack. For example, when listing containers using `docker ps` it's helpful to add the image name to the `Node` so that we can introspect what version of K8s is running on that container.

fabriziopandini replied:
thanks for the clarification

devigned commented on test/e2e/machine_pool.go:
```diff
@@ -0,0 +1,137 @@
+				ClusterName:              fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersionUpgradeFrom),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(2),
```
Certainly could. I wasn't sure if that is something that would be good to change. What I had in mind originally for Min and Max were to use those as guidelines and pick a random between `({current node count}, Max)` for scale up and then do a similar activity for Min to introduce some variability to how scale is exercised in the test suite.

fabriziopandini replied:
given that we are running e2e test in parallel on a single machine with limited resources, I think we should keep this scoped and do 0-->2 (create cluster) -->3 (scale up) -->1 (scale down) -->0 (delete)

devigned commented on test/e2e/data/infrastructure-docker/cluster-template-machine-pool.yaml:
```diff
@@ -0,0 +1,118 @@
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
```
Would you like me to move this over to using kustomize?

CecileRobertMichon replied:
since #3651 already merged, it makes sense to apply the same changes here

CecileRobertMichon commented:
/assign
@vincepri given that all review comments are addressed, can we aim to get this into 0.3.10 to unblock MachinePool E2E testing?

CecileRobertMichon commented on test/e2e/data/infrastructure-docker/cluster-template-machine-pool.yaml:
```diff
@@ -0,0 +1,118 @@
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: DockerCluster
+metadata:
+  name: '${ CLUSTER_NAME }'
```
```suggestion
  name: '${CLUSTER_NAME}'
```

CecileRobertMichon replied:
I've seen extra spaces cause issues with envsubst in the past so let's remove all extra spaces to be safe

CecileRobertMichon commented on test/e2e/config/docker.yaml:
```diff
@@ -70,25 +70,31 @@ providers:
   default/wait-cluster: ["5m", "10s"]
   default/wait-control-plane: ["10m", "10s"]
   default/wait-worker-nodes: ["5m", "10s"]
+  default/wait-machine-pool-nodes: ["5m", "10s"]
```
is this a big enough timeout? last run failed with:
```
/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/machine_pool.go:78
Timed out after 300.004s.
Expected
    <int>: 3
to equal
    <int>: 5
/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/machinepool_helpers.go:85
```

devigned replied:
I think with a reduced scale size (suggested above) we should be good.

CecileRobertMichon commented on test/infrastructure/docker/exp/doc.go:
```diff
@@ -0,0 +1,17 @@
+limitations under the License.
+*/
+
+package exp
```
what is this file doing?

devigned replied:
nothing really. removing.

CecileRobertMichon commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	cluster           *clusterv1.Cluster
+	machinePool       *clusterv1exp.MachinePool
+	dockerMachinePool *infrav1exp.DockerMachinePool
+	machines          []*docker.Machine
```
So this is using `InfraMachines` underneath, which makes sense since docker infra doesn't have a scale set concept, but maybe we should document either in code comments or in the docker exp documentation that this is a workaround for docker infra because it doesn't have a native infra resource for pool? I'm just worried about new infra providers using docker as the reference infra provider and getting misled into thinking this is how MachinePools should be implemented on the infra side.

I haven't gone through the whole PR yet so maybe this is already done.

devigned replied:
I've broken out the node pool and added docs for the type / funcs.

CecileRobertMichon reviewed (commented):
This is awesome!! 🎉

CecileRobertMichon commented:
Given that this is adding a whole new feature set to the docker provider, it should probably be a ✨ commit

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+		}
+	}
+
+	dockerMachinePool.Status.Replicas = int32(len(dockerMachinePool.Spec.ProviderIDList))
```
Should this be len (dockerMachinePool.Status.Instances)?
Otherwise, we are not counting instances not ready...

devigned replied:
You're right. I've made the change.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	dockerMachinePool.Status.Replicas = int32(len(dockerMachinePool.Spec.ProviderIDList))
+
+	if dockerMachinePool.Spec.ProviderID == "" {
+		dockerMachinePool.Spec.ProviderID = getDockerMachinePoolProviderID(cluster.Name, dockerMachinePool.Name)
```
might be worth addding a comment that this is a fake id given that in docker the is no a native abstraction grouping machiens like in Azure or other cloud provider

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	return ctrl.Result{}, nil
+}
+
+type nodePool struct {
```
Move under test/infrastructure/docker/docker so it will read docker.NodePool?

devigned replied:
Looks prettier after making the change. Great idea.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	return nil
+}
+
+func (np *nodePool) MachinesMatchingSpec() []*docker.Machine {
```
nit make private?

fabriziopandini replied:
Also, for my understanding, in this case we are concerned about the infrastructure spec (not the kubeadmconfig part).
If yes, what about 
```suggestion
func (np *nodePool) machinesMatchingInfrastructureSpec() []*docker.Machine {
```

devigned replied:
renamed

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	return matchingMachines
+}
+
+func (np *nodePool) MachinesNotMatchingSpec() []*docker.Machine {
```
nit make private?

devigned replied:
check

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	return matchingMachines
+}
+
+func (np *nodePool) AddNewMachine(ctx context.Context) error {
```
nit
```suggestion
func (np *nodePool) addMachine(ctx context.Context) error {
```

devigned replied:
done

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+
+func (np *nodePool) AddNewMachine(ctx context.Context) error {
+	instanceName := util.RandomString(6)
+	externalMachine, err := docker.NewMachine(np.cluster.Name, instanceName, np.dockerMachinePool.Spec.Template.CustomImage, map[string]string{"dmp": np.dockerMachinePool.Name}, np.logger)
```
what about making "dmp" more expressive, so if someone inspect the container it is clear the link with CAPD
e.g. `docker.cluster.x-k8s.io/machine-pool`

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+		}
+	}
+
+	return np.Refresh()
```
should this goes before the loop reconciling machines, so the new instance are reconciled immediately?

devigned replied:
Refresh is called when initializing the node pool. If the node pool were to stick around for a while, this change would probably be needed. At this point, a node pool is created in the reconcile loop and is acting on only that one docker machine pool. I don't think that much will change between construction and use of the node pool, so this probably won't have much effect.

fabriziopandini commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+		return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", machine.Name())
+	}
+
+	if err := externalMachine.PreloadLoadImages(ctx, np.dockerMachinePool.Spec.Template.PreLoadImages); err != nil {
```
am I wrong or are we going to reload images at every reconciliation?
what about doing this only if the machine is not yet bootstrapped?

fabriziopandini replied:
as alternative, we can exit earlier from this func if the machine is ready/if ProviderID is set

devigned replied:
moved inside of bootstrap conditional

fabriziopandini reviewed (commented):
@devigned 
I finally gave a look at the controller (sorry for lagging behind a little bit in reviewing this PR)

devigned commented on test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go:
```diff
@@ -0,0 +1,508 @@
+	return np, nil
+}
+
+func (np *nodePool) Refresh() error {
```
Sure.

devigned commented:
/test pull-cluster-api-e2e-full

devigned commented:
/test pull-cluster-api-e2e-full

fabriziopandini reviewed (commented):
@devigned great work!
/approve

devigned commented:
Really appreciate your insightful reviews, @fabriziopandini and @CecileRobertMichon! Thank you.

CecileRobertMichon reviewed (approved):
/lgtm

🙌

vincepri commented:
/milestone v0.3.10
